source/developer/webapp-component.rst

```diff
@@ -0,0 +1,180 @@
+
+1. Needs some data injected into it's props that the parent component doesn't have access to
+2. Needs to be able to perform some sort of action that affects the state of the store
+
+Continuing the ItemList example above, maybe our parent component doesn't care about our list of items and doesn't have access to them. Let's also imagine that we want to let the user remove items from the list by clicking on them. This means our component now needs a container for both criteria above and our props will change slightly:
+
+.. code-block:: javascript
+  :linenos:
+
+  static propTypes = {
+      /**
+       * Sets the title of the list
+       */
+      title: PropTypes.string,
+
+      /**
+       * An array of item components to display
+       */
+      items: PropTypes.arrayOf(PropTypes.object).isRequired
+
+      actions: PropTypes.shape({
+          /**
+           * An action to remove an item from the list
+           */
+          removeItem: React.PropTypes.func.isRequired
+      }).isRequired
+  }
+
+The container will then need to import selectors and actions from Redux and the connect them to the component and the store.
+
+.. code-block:: javascript
+  :linenos:
+
+  import {connect} from 'react-redux';
+  import {bindActionCreators} from 'redux';
+  import {removeItem} from 'mattermost-redux/actions/items';
+  import {getItems} from 'mattermost-redux/selectors/entities/items';
+
+  import ItemList from './item_list.js';
+
+  function mapStateToProps(state, ownProps) {
+      return {
+          ...ownProps,
+          items: getItems(state)
+      };
+  }
+
+  function mapDispatchToProps(dispatch) {
+      return {
+          actions: bindActionCreators({
+              removeItem
+          }, dispatch)
+      };
+  }
+
+  export default connect(mapStateToProps, mapDispatchToProps)(ItemList);
+
+If the selectors and/or actions you need don't yet exist in Redux then you should go add those first by following the `guide to adding actions and selectors <./redux>`__.
+
+Your ``index.js`` and ``item_list.js`` files will live together in an ``item_list/`` directory.
+
+Implementing your Component
+------------------
+
+With the props defined and, if necessary, the container built, you're ready to implement your component. For the most part implementing a component for the webapp is no different than building any other React component. That said, we do have a few rules to abide by:
+
+1. All components must extend ``React.PureComponent``
+2. All data for rendering must come from props
+3. Local rendering information that does not affect the store may be kept in state
+
+Our ItemList example might look something like this:
+
+.. code-block:: javascript
+  :linenos:
+
+  export default class ItemList extends React.PureComponent {
```

Comment:
Okay, I'll go back to that on the web app then. It'll save us an import, especially when prop types becomes its own library